src/js/panels.js

```diff
@@ -198,16 +193,377 @@ https://github.com/fluid-project/infusion/raw/master/Infusion-LICENSE.txt
+            }
+        },
+        components: {
+            attachTooltipOnLang: {
+                type: "gpii.firstDiscovery.panel.lang.attachTooltipOnLang",
+                container: "{lang}.container",
+                options: {
+                    selectors: "{lang}.options.selectors",
+                    modelRelay: {
+                        source: "{lang}.model.lang",
+                        target: "currentSelectedIndex",
+                        singleTransform: {
+                            type: "fluid.transforms.free",
+                            args: {
+                                "langs": "{lang}.options.controlValues.lang",
+                                "currentLang": "{lang}.model.lang"
+                            },
+                            func: "gpii.firstDiscovery.panel.lang.getCurrentSelectedIndex"
+                        }
+                    },
+                    tooltipContentMap: {
+                        "prev": "navButtonTooltip",
+                        "next": "navButtonTooltip",
+                        "langRow": {
+                            tooltip: "{lang}.options.stringArrayIndex.tooltip",
+                            tooltipAtSelect: "{lang}.options.stringArrayIndex.tooltipAtSelect"
+                        },
+                        "langInput": {
+                            tooltip: "{lang}.options.stringArrayIndex.tooltip",
+                            tooltipAtSelect: "{lang}.options.stringArrayIndex.tooltipAtSelect"
+                        }
+                    },
+                    listeners: {
+                        "{lang}.events.afterRender": {
+                            funcName: "{that}.tooltip.applier.change",
+                            args: ["idToContent", {
+                                expander: {
+                                    func: "{that}.tooltip.getTooltipModel"
+                                }
+                            }]
+                        }
+                    }
+                }
+            }
+        },
+        model: {
+            firstLangSelected: false,
+            lastLangSelected: false
+        },
+        modelRelay: [{
+            target: "firstLangSelected",
+            singleTransform: {
+                type: "fluid.transforms.binaryOp",
+                left: "{that}.model.lang",
+                operator: "===",
+                right: {
+                    expander: {
+                        funcName: "fluid.get",
+                        args: ["{that}.options.controlValues.lang", 0]
+                    }
+                }
+            }
+        }, {
+            target: "lastLangSelected",
+            singleTransform: {
+                type: "fluid.transforms.binaryOp",
+                left: "{that}.model.lang",
+                operator: "===",
+                right: {
+                    expander: {
+                        funcName: "gpii.firstDiscovery.panel.lang.getLastArrayElement",
```

Comment:
Both of these relays should be replaced by a sequence of two relays - firstly, a relay between the language selected, and the index of it - using the indexOf transform described above. Secondly, we can have relays that imply compare the index with 0 and {that}.options.controlValues.lang.length. We can eliminate the "getLastArrayElement" function entirely.